src/go/rpk/pkg/cli/cmd/debug/bundle/bundle_linux.go

```diff
@@ -743,23 +743,144 @@ func saveDmidecode(ctx context.Context, ps *stepParams) step {
+
+// walkSizeDir walks the directory tree rooted at the given path and calculates
+// the size of each file in bytes. It also excludes files whose names match
+// the given regular expression 'exclude'.
+//
+// The function returns a slice of fileSize struct, each of which contains the
+// path and size of a file, as well as the total size of the directory in bytes.
+func walkSizeDir(dir string, exclude *regexp.Regexp) (files []fileSize, size int64, err error) {
+	err = filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if !info.IsDir() {
+			if !exclude.MatchString(info.Name()) {
+				size += info.Size()
+				files = append(files, fileSize{
+					path: path,
+					size: info.Size(),
+				})
+			}
+		}
+		return nil
+	})
+	return files, size, err
+}
+
+// sortControllerLogDir takes a slice of fileSize structs that represents the
+// controller logs directory and sorts it by the base_offset and term integers
+// in the filenames. Filenames should follow this format:
+// {base_offset}-{term}-{version}. If multiple files have the same base_offset,
+// then the function will sort them based on their term.
+func sortControllerLogDir(dir *[]fileSize) {
+	sort.Slice(*dir, func(i, j int) bool {
+		filename1 := filepath.Base((*dir)[i].path)
+		filename2 := filepath.Base((*dir)[j].path)
+
+		f1 := strings.Split(filename1, "-")
+		f2 := strings.Split(filename2, "-")
+
+		// One of the filenames don't follow {base_offset}-{term}-{version}.
+		if len(f1) < 3 || len(f2) < 3 {
+			return false
+		}
+
+		// Here, we parse the base_offset. Any errors can be safely ignored
+		// because if it is a controller log, the string will be parsed
+		// correctly, otherwise, the value will be 0 and the slice won't be
+		// sorted.
+		offset1, _ := strconv.Atoi(f1[0])
+		offset2, _ := strconv.Atoi(f2[0])
+
+		// If the base offsets are different, sort by base_offset.
+		if offset1 != offset2 {
+			return offset1 < offset2
+		}
+
+		// If they are the same, we sort based on the term
+		term1, _ := strconv.Atoi(f1[0])
+		term2, _ := strconv.Atoi(f2[0])
+
+		return term1 < term2
+	})
+}
+
+// sliceControllerDir takes a slice of fileSize structs and a byte size limit
+// (logLimitBytes). It returns a slice with the files that fit within the limit,
+// copied from both the head and tail of the input slice.
+func sliceControllerDir(cFiles []fileSize, logLimitBytes int64) (slice []fileSize) {
+	// We start copying the files from the head until we reach the first half of
+	// the limit:
+	var headAcc int64
+	half := logLimitBytes / 2
+	for i := 0; i < len(cFiles); i++ {
+		cLog := cFiles[i]
+		if headAcc+cLog.size > half {
+			break
+		}
+		slice = append(slice, cLog)
+		headAcc += cLog.size
+	}
+
+	// Now from the tail until we fill the remaining bytes:
+	var tailAcc int64
+	// We don't use half since headAcc could be < than half.
+	remainingBytes := logLimitBytes - headAcc
+	for i := len(cFiles) - 1; i > 0; i-- {
```

Comment:
`for i, alreadyTaken := len(cFiles) - 1, len(slice); i > alreadyTaken; i-- `

so that this does not take a controller log twice

Reply:
Although, this condition is guarded by line R859 below `if int(size) < logLimitBytes` -- a file will not be taken twice here because, if it would be taken twice, that means the total allowed size is larger than the directory size and we take the shortcut -- I still think it's worth it to make the logic _here_ correct from a paranoid standpoint

Reply:
Yes, exactly. But good point, better to be paranoid here. :+1: